<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Title of your project - 5&nbsp; Working on GenomeDK</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../howtos/reproducible.html" rel="next">
<link href="../notebooks/02_notebook.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../howtos/genomedk.html">Howtos</a></li><li class="breadcrumb-item"><a href="../howtos/genomedk.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Working on GenomeDK</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Title of your project</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Outline</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Reports</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../reports/manuscript.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Manuscript</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../reports/supplementary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Supplementary info</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Notebooks</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/01_notebook.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Some title</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/02_notebook.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Some title</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Howtos</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../howtos/genomedk.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Working on GenomeDK</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../howtos/reproducible.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Reproducible research</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul class="collapse">
  <li><a href="#setting-up-your-own-machine" id="toc-setting-up-your-own-machine" class="nav-link active" data-scroll-target="#setting-up-your-own-machine"><span class="header-section-number">5.1</span> Setting up your own machine</a></li>
  <li><a href="#setting-up-your-home-on-the-cluster" id="toc-setting-up-your-home-on-the-cluster" class="nav-link" data-scroll-target="#setting-up-your-home-on-the-cluster"><span class="header-section-number">5.2</span> Setting up your home on the cluster</a></li>
  <li><a href="#backup-and-version-control" id="toc-backup-and-version-control" class="nav-link" data-scroll-target="#backup-and-version-control"><span class="header-section-number">5.3</span> Backup and version control</a></li>
  <li><a href="#working-on-the-cluster" id="toc-working-on-the-cluster" class="nav-link" data-scroll-target="#working-on-the-cluster"><span class="header-section-number">5.4</span> Working on the cluster</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Working on GenomeDK</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><img src="https://icon-library.com/images/62418-database-icons-virtual-servers-computer-private-server.png" align="right" width="200"></p>
<p>This repository serves as a walk-through to begin a data analysis project on the <a href="https://genome.au.dk/">GenomeDK cluster</a>. Be that a PiB, BSc, MSc, or Ph.D.&nbsp;project. For more seasoned users, it can serve as a showcase of best practices. Some basic experience executing commands in a “terminal” is assumed.</p>
<p>The cluster is a very large collection of computers with a shared file system. It does not have a screen and a keyboard you can go use. However, by connecting to the cluster from your computer, you can create and edit files much like if they were on your machine. The goal here is to take you through all the steps required to make this possible.</p>
<p><strong>NB:</strong> This is not a leisure read. Make sure you follow it to the letter, doing each step in order.</p>
<p>Below you will see something like <code>&lt;cluster user name&gt;</code> or <code>&lt;project folder&gt;</code> a lot. When ever you see something like that you should replace the whole thing, including <code>&lt;</code> and <code>&gt;</code>, with whatever is says. E.g., if your cluster user name is <code>preben</code> you should replace <code>&lt;cluster user name&gt;</code> with <code>preben</code>.</p>
<p>Before you can begin, you need access to the cluster. The cluster is called GenomeDK and has its own <a href="https://genome.au.dk">website</a> with lots of information and documentation. To get an account on the cluster, you must <a href="https://genome.au.dk/docs/getting-started/#request-access">request one here</a>. Below, <code>username</code> will represent your user name.</p>
<p>On the cluster, you have a home folder that only you have can access. That is where you end up when you log in. Collaborative projects or projects that use or generate a lot of data projects belong in project folders. If you do a project, we will set up a dedicated project folder for this.</p>
<p>Your project is a research project. Even if it is meant for training, think of it as training reproducibility too. Your efforts, experiences, results, and developed expertise is valuable to all that comes after you. So do not make your project a dead end; make it a shoulder to stand on.</p>
<p>Reproducibility basically means that anyone should be able to do <em>exactly</em> what you did to <em>exactly</em> reproduce your results. They should be able to get the same results in terms of numbers, plots, etc. This requires care and attention to detail, but it is not difficult. This page will help you get set up to do reproducible research.</p>
<section id="setting-up-your-own-machine" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="setting-up-your-own-machine"><span class="header-section-number">5.1</span> Setting up your own machine</h2>
<p>Before we get to the cluster, we need to get you properly set up on your own machine.</p>
<section id="install-python" class="level3" data-number="5.1.1">
<h3 data-number="5.1.1" class="anchored" data-anchor-id="install-python"><span class="header-section-number">5.1.1</span> Install Python</h3>
<p>If you have not done so already, you should install a distribution of Python called <em>Anaconda</em>. Anaconda is not only an easy way of installing Python on Windows, Mac, and Linux; it also comes with the conda package management system (see below). To install Anaconda, head to <a href="https://www.anaconda.com/download">this page</a>. When the download has been completed, you must follow the default installation.</p>
</section>
<section id="the-terminal" class="level3" data-number="5.1.2">
<h3 data-number="5.1.2" class="anchored" data-anchor-id="the-terminal"><span class="header-section-number">5.1.2</span> The Terminal</h3>
<p>Most of the programs we will use in this course are command-line applications. I.e., programs that are executed by writing their name and any arguments in a terminal rather than clicking on an icon and using a graphical user interface. There are many different programs that can serve as a terminal. If you have a Windows machine, you must use the <em>Anaconda Powershell Prompt</em> (<em>not</em> the Anaconda Prompt and <em>not</em> the <code>CMD</code>). You installed Anaconda Powershell Prompt along with Anaconda Python. If you have a Mac, the terminal you will use is called <em>Terminal</em>. The Terminal application is pre-installed on Mac. So from now on, whenever I refer to the terminal, I mean <em>Anaconda Powershell Prompt</em> on Windows and <em>Terminal</em> on Mac. We will assume some familiarity with using a terminal and with executing commands on the command line. If you have not used a terminal before, or if you are a bit rusty, you should run through <a href="https://lifehacker.com/5633909/who-needs-a-mouse-learn-to-use-the-command-line-for-almost-anything">this primer</a> before you go on.</p>
</section>
<section id="conda-environments" class="level3" data-number="5.1.3">
<h3 data-number="5.1.3" class="anchored" data-anchor-id="conda-environments"><span class="header-section-number">5.1.3</span> Conda environments</h3>
<p><img src="https://docs.conda.io/en/latest/_images/conda_logo.svg" align="right" width="200" height="100"></p>
<p>You need to install packages and programs for use in your analyses and pipelines. Sometimes, however, the versions of packages you need for one project conflicts with the versions you need for other projects that you work on in parallel. Such conflicts seem like an unsolvable problem. Would it not be fantastic if you could create a small world insulated from the rest of your Anaconda installation. Then that small world would only contain the packages you needed for a single project. If each project had its own isolated world, then there would be no such conflicts. Fortunately, there is a tool that lets you do just that, and its name is Conda. The small worlds that Conda creates are called “environments,” and you can create as many as you like. You can then switch between them as you switch between your bioinformatics projects. Conda also downloads and installs the packages for you, and it makes sure that the packages you install in each environment are compatible. It even makes sure that packages needed by packages (dependencies) are installed too. By creating an environment for each project, the libraries installed for each project do not interfere.</p>
</section>
<section id="create-an-environment-on-your-local-machine" class="level3" data-number="5.1.4">
<h3 data-number="5.1.4" class="anchored" data-anchor-id="create-an-environment-on-your-local-machine"><span class="header-section-number">5.1.4</span> Create an environment on your local machine</h3>
<p>When you install Anaconda or Miniconda, Conda makes a single base environment for you. It is called “base,” and this is why it says “(base)” at your terminal prompt. You need a conda environment for your project on both your local machine and on the cluster. Let us call both of them ‘bircproject’ (you can call it anything you like).</p>
<p>The environment on your local machine does not need a lot of packages since it mainly serves to let you connect to the cluster. This creates the environment and installs <code>slurm-jupyter</code> from my conda channel:</p>
<pre><code>conda create -n bircproject -c kaspermunch slurm-jupyter</code></pre>
<p>Say yes (press Enter) when asked to install packages.</p>
</section>
<section id="vpn" class="level3" data-number="5.1.5">
<h3 data-number="5.1.5" class="anchored" data-anchor-id="vpn"><span class="header-section-number">5.1.5</span> VPN</h3>
<p>To be able to connect to the cluster, you need to on the AU internal network. You can do that by either physically being on campus or by connecting to the AU network using VPN. To install VPN, use the instructions <a href="https://studerende.au.dk/it-support/vpn/">on this page</a>. Before you can <em>use</em> the VPN, you also need to enable two-step verification. You can see how to do that on the same page. If you are not physically on campus, you need to activate your VPN before you can log in to the cluster. Your password for VPN is the same as you use to log on to access Blackboard.</p>
</section>
<section id="connecting-to-the-cluster-using-ssh" class="level3" data-number="5.1.6">
<h3 data-number="5.1.6" class="anchored" data-anchor-id="connecting-to-the-cluster-using-ssh"><span class="header-section-number">5.1.6</span> Connecting to the cluster using ssh</h3>
<p><img src="https://atulhost.com/wp-content/uploads/2020/04/ssh.png" align="right" width="300" height="200"></p>
<p>SSH is short for secure shell. A shell is the software that lets you run commands in your terminal window. The <em>secure shell</em> (SSH) lets you securely log in to another computer so you can navigate the folders and run commands on that machine. So when you open your terminal window, your commands run on your local machine, but when you “ssh” (yes it is a verb too) into the cluster, your commands now run on the cluster. Before you go on, try to run the command <code>hostname</code> in you terminal. You can see that it prints something that tells you that you are on your own computer.</p>
<p>You connect to the cluster from the terminal by executing this command (remember to replace <code>&lt;cluster user name&gt;</code> with your actual cluster user name):</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ssh</span> <span class="op">&lt;</span>cluster user name<span class="op">&gt;</span>@login.genome.au.dk</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>When you do, you are prompted for the password for your cluster username. Enter that and press enter. You are now in your home folder on the cluster. Your terminal looks the same as before, but it will print:</p>
<pre><code>  _____                                ______ _   __
 |  __ \                               |  _  \ | / /
 | |  \/ ___ _ __   ___  _ __ ___   ___| | | | |/ /
 | | __ / _ \ '_ \ / _ \| '_ ` _ \ / _ \ | | |    \
 | |_\ \  __/ | | | (_) | | | | | |  __/ |/ /| |\  \
  \____/\___|_| |_|\___/|_| |_| |_|\___|___/ \_| \_/</code></pre>
<p>If you run the <code>hostname</code> command again, you can see that you are on <code>fe-open-01</code>. Now log out of the cluster again. You do that using the <code>exit</code> command or by pressing <code>Ctrl-d</code>. Now you are back on your own machine. Try <code>hostname</code> again and see what your own machine is called.</p>
<p>You will need to log in to the cluster many many times, so you should set up your <code>ssh</code> connection to the cluster so you can connect securely without typing the password every time. You do not need to know <em>how</em> this works, but if you are interested, here is roughly how:</p>
<blockquote class="blockquote">
<p>Firstly, you have to understand what public/private encryption keys are. A private key is a very long, random sequence of bits. A private key is kept secret and never leaves your own machine. A public key is another string of bits that is a derivative of the private key. You can generate a unique public key from the private key but cannot get the private key from a public key: It is a one-way process. Using the public key, you can encrypt (or sign) any message, and it will only be possible to decrypt it using the private key. In other words, anyone with your public key can send you encrypted messages that only you will be able to read. So, if the cluster has your public key saved, it can authenticate you like this: The cluster sends your machine a message that is encrypted using your public key. Your machine then decrypts the message using its private key and sends it back. If the cluster agrees it is decrypted correctly, it logs you in.</p>
</blockquote>
<p>First, check if you have these two authentication files on your local machine (you can do so by running <code>ls -a ~/.ssh</code> in the terminal):</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/.ssh/id_rsa</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">~/.ssh/id_rsa.pub</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You most likely do not. If so, you generate a pair of authentication keys with the command below. Just press Enter when asked “Enter file in which to save the key”. Do not enter a passphrase when prompted - just press enter:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ssh-keygen</span> <span class="at">-t</span> rsa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now use <code>ssh</code> to create a directory <code>~/.ssh</code> on the cluster (assuming your username on the cluster is <code>&lt;cluster user name&gt;</code>). You will be prompted for your password.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ssh</span> <span class="op">&lt;</span>cluster user name<span class="op">&gt;</span>@login.genome.au.dk mkdir <span class="at">-p</span> .ssh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Finally, append the public <code>ssh</code> key on your local machine to the file <code>.ssh/authorized_keys</code> on the cluster and enter your password (replace <code>&lt;cluster user name&gt;</code> with your cluster user name):</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> ~/.ssh/id_rsa.pub <span class="kw">|</span> <span class="fu">ssh</span> username@login.genome.au.dk <span class="st">'cat &gt;&gt; .ssh/authorized_keys'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>From now on, you can log into the cluster from your local machine without being prompted for a password.</p>
</section>
</section>
<section id="setting-up-your-home-on-the-cluster" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="setting-up-your-home-on-the-cluster"><span class="header-section-number">5.2</span> Setting up your home on the cluster</h2>
<p>Now log in to the cluster:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ssh</span> <span class="op">&lt;</span>cluster user name<span class="op">&gt;</span>@login.genome.au.dk</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>(see, no password).</p>
<section id="install-python-on-your-cluster-account" class="level3" data-number="5.2.1">
<h3 data-number="5.2.1" class="anchored" data-anchor-id="install-python-on-your-cluster-account"><span class="header-section-number">5.2.1</span> Install Python on your cluster account</h3>
<p>You need to install miniconda (a minimal Anaconda version) in your cluster home dir. Log in to the cluster and run this command to download the miniconda install script:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Then use this command to download and install miniconda:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bash</span> Miniconda3-latest-Linux-x86_64.sh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Follow the default installation, and say <strong>yes</strong> when it asks you if it should run <code>conda init</code> for you.</p>
<p><strong>NP:</strong> Now log out of the cluster and log back in. This is needed to make the <code>conda</code> command available to you.</p>
</section>
<section id="creating-an-environment-on-the-cluster" class="level3" data-number="5.2.2">
<h3 data-number="5.2.2" class="anchored" data-anchor-id="creating-an-environment-on-the-cluster"><span class="header-section-number">5.2.2</span> Creating an environment on the cluster</h3>
<p>Log in to the cluster and run this command to create a conda environment for your project on the cluster. This environment should contain the packages that you need for your project. Unless your supervise tells you otherwise, the command below should install what you need to get started on a project:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">--name</span> bircproject <span class="at">-c</span> gwforg <span class="at">-c</span> conda-forge <span class="at">-c</span> bioconda <span class="at">-c</span> kaspermunch slurm-jupyter gwf jupyterlab jupyterlab-git pandas seaborn biopython </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>If your supervisor wants you to install a specific set of packages for your project, he/she may have given you a file called <code>environment.yml</code>. In that case you create you instead create your environment like this:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> env create <span class="at">-n</span> bircproject <span class="at">-f</span> environment.yml</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You probably end up needing more packages than you initially included, you can easily install them later. E.g., to see how to install <code>biopython</code> using <code>conda</code>, just google “conda biopython”. The top link instructs you to install it like this: <code>conda install -c conda-forge biopython</code>.</p>
<p><strong>Important:</strong> Whenever you log into the cluster to work on your project, you should activate your <code>bircproject</code> environment like this:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bircproject</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>When your environment is active, it says <code>(bircproject)</code> on the command prompt instead of <code>(base)</code>.</p>
</section>
<section id="set-up-jupyter" class="level3" data-number="5.2.3">
<h3 data-number="5.2.3" class="anchored" data-anchor-id="set-up-jupyter"><span class="header-section-number">5.2.3</span> Set up Jupyter</h3>
<p><img src="https://github.com/jupyter/jupyter.github.io/blob/master/assets/main-logo.svg" align="right" width="100" height="100"></p>
<p><a href="https://jupyter.org/">Jupyter</a> is a notebook environment where you can easily combine text, code, and plots. Using the <a href="https://slurm-jupyter.readthedocs.io/en/latest">slurm-jupyter</a> tool, you can run a jupyter notebook on the cluster but see it in the browser on your own machine. That way, your analysis runs on the cluster file system where your data is but the notebook interface is sent to your browser window.</p>
<p>Before you can connect to a jupyter session on the cluster, you need to do a bit of configuration of the jupyter settings on the cluster. <code>slurm-jupyter</code> comes with script that automates that procedure. Just log into the cluster, activate your environment, and run:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">config-slurm-jupyter.sh</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The script will ask for a lot of information. You can just press <code>Enter</code> for all of them <strong>except</strong> when prompted for what password you want to use: Then, you must type your cluster password.</p>
</section>
</section>
<section id="backup-and-version-control" class="level2" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="backup-and-version-control"><span class="header-section-number">5.3</span> Backup and version control</h2>
<p>Your files on the cluster are not backed up! If you want to backup files, you need to put them in a folder called BACKUP. However, a better way is to use git and GitHub (see below).</p>
<section id="github" class="level3" data-number="5.3.1">
<h3 data-number="5.3.1" class="anchored" data-anchor-id="github"><span class="header-section-number">5.3.1</span> GitHub</h3>
<p><img src="https://github.githubassets.com/images/modules/open_graph/github-mark.png" align="right" width="200"></p>
<p>The page you are reading now is a GitHub repository. In addition to the documentation you are currently reading, it serves as a template for your project.</p>
<p>Start by creating your own <a href="https://github.com/join">github account</a> if you do not have one.</p>
<p>Then you need to follow the instructions <a href="https://www.inmotionhosting.com/support/server/ssh/how-to-add-ssh-keys-to-your-github-account/">on this page</a> to add ssh keys to GitHub.</p>
<p>Now log in to your GitHub account on the web. Go to the <a href="https://github.com/kaspermunch/birc-project">birc-project repository</a>. In addition to this page, the repository contains a folder structure and config files, which serve as a good starting point for a project repository. You must now “fork” the repository. Forking creates your own copy of the <code>birc-project</code> repository under your own GitHub account. To fork the repository, click the button in the top right corner that says <code>Fork</code>.</p>
<p>Once your own <code>birc-project</code> repository is ready, go to the repository front page. There you will find your own copy of this page <code>README.md</code>. Read on from there.</p>
</section>
<section id="git" class="level3" data-number="5.3.2">
<h3 data-number="5.3.2" class="anchored" data-anchor-id="git"><span class="header-section-number">5.3.2</span> Git</h3>
<p><a href="https://git-scm.com/">Git</a> is a version control tool that you use from the terminal. A folder under Git control is called a repository. <code>birc-project</code> is such a repository. Git does not interfere with your files and it does not save them. It lets you monitor the state of your files so you can easily see if any files are added, modified, or removed, and it allows you to (manually) maintain a record of what files where changes when, how, and for what reason.</p>
<p>Start logging into the cluster and run these two commands to let Git know who you are:</p>
<pre><code>git config --global user.name "&lt;Your GitHub user name&gt;"
git config --global user.email &lt;your_email@whatever.com&gt;</code></pre>
<p>Note that to run git commands tracking your <code>birc-project</code>, you must stand in your <code>birc-project</code> folder when you run them.</p>
<p><a href="https://www.atlassian.com/git">These tutorials</a> are good. The three most important commands to learn are <a href="https://git-scm.com/docs/git-status">git status</a>, <a href="https://git-scm.com/docs/git-add">git add</a>, <a href="https://git-scm.com/docs/git-rm">git rm</a>, <a href="https://git-scm.com/docs/git-commit">git commit</a> and <a href="https://git-scm.com/docs/git-push">git push</a>.</p>
<p>You should read the documentation, but in a nutshell this is how it works:</p>
<p><code>git status</code> shows you the status of your files. It may show you that you have created or modified one or more files.</p>
<p>To journal the creation or modification of a file, you run <code>git add &lt;the_changed_file&gt;</code>. This “stages” the change. This adds the file to the a “group” of changes that represent some modification to your project. You can add more files to that “group” by running <code>git add</code> again. if Maybe you added two python files that together lets you run some analysis. Now you want to store the group of changes to the journal maintained by Git and associate it with a description that describes it. For that you use <code>git commit -m 'description'</code>. If you added two python files, your description could be <code>'Added two python files for my analysis'</code>. Now your changes are recorded in the Git journal for the <code>birc-project</code> repository on the cluster, but that does not serve as a backup in case you accidentally delete your entire folder or the cluster burns down. To backup your repository you need to synchronize your local <code>birc-project</code> repository on your computer with the <code>birc-project</code> repository on GitHub. You can see how below.</p>
<p>There is a cheat sheet <a href="https://training.github.com/downloads/github-git-cheat-sheet/">here</a> and some good visual guides <a href="https://ndpsoftware.com/git-cheatsheet.html">here</a> and <a href="https://marklodato.github.io/visual-git-guide/index-en.html">here</a>.</p>
<p>Rember: Any files you push to GitHub are backed up. Make a habit of pushing all changes you make.</p>
</section>
</section>
<section id="working-on-the-cluster" class="level2" data-number="5.4">
<h2 data-number="5.4" class="anchored" data-anchor-id="working-on-the-cluster"><span class="header-section-number">5.4</span> Working on the cluster</h2>
<section id="the-project-folder" class="level3" data-number="5.4.1">
<h3 data-number="5.4.1" class="anchored" data-anchor-id="the-project-folder"><span class="header-section-number">5.4.1</span> The project folder</h3>
<p>The project folder is a folder that is set up on the cluster to hold your project. I call it <code>projectfolder</code> here, but it will be called something sensible like <code>baboonadmixture</code>.</p>
<p>It is accessible to only you and anyone else you collaborate with (such as your supervisor). The project folder is in your home directory and should hold the following subfolders:</p>
<pre><code>projectfolder
    /data
    /people
        /username
        /supervisor_username</code></pre>
<p>The <code>&lt;projectfolder&gt;/people/&lt;username&gt;</code> is your domain. This is where you have all the files that relates to your project.</p>
<p>The name of the project folder is is also the name of the account that you should bill any work on the cluster to. When you run <code>srun</code>, <code>sbatch</code> or <code>slurm-jupyter</code> you myst specify that project name using the <code>-A</code> or <code>--account</code> options (see below for more details on that).</p>
</section>
<section id="cloning-your-git-repository-to-the-cluster" class="level3" data-number="5.4.2">
<h3 data-number="5.4.2" class="anchored" data-anchor-id="cloning-your-git-repository-to-the-cluster"><span class="header-section-number">5.4.2</span> Cloning your git repository to the cluster</h3>
<p>Go to your folder under the project folder (<code>&lt;projectfolder&gt;/people/&lt;username&gt;</code>) and clone your repository:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone git@github.com:username/birc-project.git</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>(replace <code>username</code> with your GitHub username).</p>
<p>You now have a folder called <code>&lt;projectfolder&gt;/people/&lt;username&gt;/birc-project</code> and this is where you will do all your work for the project.</p>
<p>If you <code>cd</code> into <code>birc-directory</code> and run <code>ls</code>, you will see a number of folders.</p>
<ul>
<li><code>data</code>: stores <em>small</em> (tens of megabases) data files you want to keep .</li>
<li><code>docs</code>: stores text files explaining how you set up your project and analyzes.</li>
<li><code>binder</code>: stores an <code>environment.yml</code> files documenting your conda environment used in the project.</li>
<li><code>sandbox</code>: stores experiments that are not yet part of your project workflow.</li>
<li><code>scripts</code>: stores Python scripts that that produces intermediate and final results.</li>
<li><code>steps</code>: stores intermediary files (“steps” on the way to final results).</li>
<li><code>notebooks</code>: stores Juptyer notebooks with code, documentation, and results.</li>
<li><code>figures</code>: stores result plots and figures you make.</li>
<li><code>results</code>: stores the small result files of your project (tens of megabases).</li>
</ul>
<p>Files in all those folders are under Git control, <em>except</em> files in the <code>steps</code> folder. Those files are not backed up in any way, but should instead be reproducible using the code and information in your other folders.</p>
</section>
<section id="visual-studio-code" class="level3" data-number="5.4.3">
<h3 data-number="5.4.3" class="anchored" data-anchor-id="visual-studio-code"><span class="header-section-number">5.4.3</span> Visual Studio Code</h3>
<p><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/9a/Visual_Studio_Code_1.35_icon.svg/2560px-Visual_Studio_Code_1.35_icon.svg.png" align="right" width="100" height="100"></p>
<p>If you did not do so when you installed Anaconda, you should download and install Visual Studio Code. VScode is great for developing scripts and editing text files. Once you have installed VS code, you should install the “Remote Development” extension. You do that by clicking the funny squares in the left bar and search for “Remote Development”. Once installed, you can click the small green square in the lower-left corner to connect to the cluster. Select “Connect current window to host” then “Add new SSH host”, then type <code>&lt;username&gt;@login.genome.au.dk</code>, then select the config file <code>.ssh/config</code>. Now you can click the small green square in the lower-left corner to connect to the cluster by selecting <code>login.genome.au.dk</code>. It may take a bit, but once it is done installing a remote server, you will have access to the files in your home folder on the cluster.</p>
</section>
<section id="how-to-run-a-jupyter-notebook-on-the-cluster" class="level3" data-number="5.4.4">
<h3 data-number="5.4.4" class="anchored" data-anchor-id="how-to-run-a-jupyter-notebook-on-the-cluster"><span class="header-section-number">5.4.4</span> How to run a Jupyter notebook on the cluster</h3>
<p>Jupyter runs best in the <a href="https://www.google.com/chrome">Chrome browser</a> or Safari on Mac. For the best experience, install that before you go on. It does not need to be your default browser. <code>slurm-jupyter</code> will use it anyway. Now make sure you are on your own machine and that your <code>popgen</code> environment is activated. Then run this command to start a jupyter notebook on the cluster and send the display to your browser:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">slurm-jupyter</span> <span class="at">-C</span> <span class="at">-u</span> <span class="op">&lt;</span>cluster_user_name<span class="op">&gt;</span> -A <span class="op">&lt;</span>projectfolder<span class="op">&lt;&gt;</span> -e bircproject</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>(replace <code>&lt;cluster_user_name&gt;</code> with your cluster user name, <code>&lt;projectfolder&gt;</code> with your project folder name.</p>
<p>Watch the terminal to see what is going on. After a while, a jupyter notebook should show up in your browser window. The first time you do this, your browser may refuse to show jupyter because the connection is unsafe. In Safari you are prompted with this winidow where you click “details”:</p>
<p><img src="img/mac_warning1.png" alt="image" width="450"></p>
<p>Then you get this window and click “visit this website”:</p>
<p><img src="img/mac_warning2.png" alt="image" width="450"></p>
<p>In Chrome, you can simply type the characters “thisisunsafe” while in the Chrome window:</p>
<p><img src="img/thisisunsafe.png" alt="image" width="450"></p>
<p>Once ready, jupyter may ask for your cluster password. To close the jupyter notebook, press <code>Ctrl-c</code> in the terminal. Closing the browser window does <strong>not</strong> close down the jupyter on the cluster. You can <a href="https://www.dataquest.io/blog/jupyter-notebook-tutorial/">read this tutorial</a> to learn how to use a jupyter notebook.</p>
</section>
<section id="running-interactive-commands-on-the-cluster" class="level3" data-number="5.4.5">
<h3 data-number="5.4.5" class="anchored" data-anchor-id="running-interactive-commands-on-the-cluster"><span class="header-section-number">5.4.5</span> Running interactive commands on the cluster</h3>
<p>When you log into the cluster, you land on the “front-end” of the cluster. Think of it as the lobby of a giant hotel. If you execute the <code>hostname</code> command, you will get <code>fe-open-01</code>. <code>fe1</code> is the name of the front-end machine. The “front-end” is a single machine shared by anyone who logs in. So you cannot run resource-intensive jobs there, but quick commands are ok. Commands that finish in less than ten seconds are ok. In the exercises for this course, you will run software that takes a much longer time to finish. So you need one of the computing machines on the cluster, so you can work on that instead. You ask for a computing machine by running this command:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--mem-per-cpu</span><span class="op">=</span>1g <span class="at">--time</span><span class="op">=</span>3:00:00 <span class="at">--account</span><span class="op">=&lt;</span>projectfolder<span class="op">&gt;</span> --pty bash</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>That says that you need at most one gigabyte of memory, that you need it for at most three hours (the duration of the exercise), and that the computing expenses should be billed to the project <code>&lt;projectfolder&gt;</code>. When you execute the command, your terminal will say “srun: job 40924828 queued and waiting for resources”. That means that you are waiting for a machine. Once it prints “srun: job 40924828 has been allocated resources”, you have been logged into a computing node. If you execute the <code>hostname</code> command, you will get something like <code>s05n20</code>. <code>s05n20</code> is a computing machine. The same way you moved from your own computer to front-end machine of the cluster by logging in using ssh, the command above moves you from the front-end to a compute machine. Now you can execute any command you like without causing trouble for anyone.</p>
<p>Now try to log out of the compute node by executing the <code>exit</code> command or by pressing <code>Ctrl-d</code>. If you execute the <code>hostname</code> command again, you will get <code>fe1.genomedk.net</code> showing that you are back at the front-end machine.</p>
</section>
<section id="queueing-commands-on-the-cluster" class="level3" data-number="5.4.6">
<h3 data-number="5.4.6" class="anchored" data-anchor-id="queueing-commands-on-the-cluster"><span class="header-section-number">5.4.6</span> Queueing commands on the cluster</h3>
<p>For non-interactive work, it is better to submit your command as a job to the cluster. When you do that, the job gets queued along with many other jobs, and as soon as the requested resources are available on the cluster, the job will start on one the many many machines. To submit a job, you must first create a file (a “batch script”) that contains both the requested computer resources and the command you want to run.</p>
<p>Create a file called <code>myscript.sh</code> with exactly this content:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mem=1gb</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --time=01:00:00</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --account=&lt;projectfolder&gt;</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --job-name=firstjob</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"I can submit cluster jobs now!"</span> <span class="op">&gt;</span> success.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>(replace <code>&lt;projectfolder&gt;</code> with your project folder name)</p>
<p>The first line says this is a bash script, the lines following three lines say that your job needs at most one gigabyte of memory, will run for at most one hour, that the expenses should be billed to the project <code>&lt;projectfolder&gt;</code>. The fourth line gives the name of the job. Here we have called it <code>firstjob</code>, but you should name it something sensible.</p>
<p>You submit the job using the <code>sbatch</code> command:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sbatch</span> myscript.sh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now your job is queued. Use the <code>mj</code> command to see what jobs you have queued or running. That will show something like this:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode txt code-with-copy"><code class="sourceCode default"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>                                                                        Alloc</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>Job ID           Username Queue    Jobname    SessID NDS  S Elap Time   nodes</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>---------------- -------- -------- ---------- ------ ---  - ----------  -----</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>34745986         kmt      normal   firstjob       --   1  R 0-00:19:27  s03n56</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>If you want to cancel this job before it finishes, you can use the <code>scancel</code> command:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scancel</span> 34745986</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Once your job finishes, it has created the file <code>success.txt</code> and written “I can submit cluster jobs now!” to it. So see that you can use the <code>cat</code> command:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> success.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>When you a program or script on the command line, it usually also prints some information in the terminal. When you run a job on the cluster there is no terminal to print to. Instead, this is written to two files that you can read when the job finishes. In this case, the fiels are called <code>firstjob.stdout</code> and <code>firstjob.stderr</code>. To see what is in them, you can use the <code>cat</code> command:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> firstjob.stdout</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>and</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> firstjob.stderr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>That is basically it.</p>
</section>
<section id="how-to-copy-files-to-and-from-the-cluster" class="level3" data-number="5.4.7">
<h3 data-number="5.4.7" class="anchored" data-anchor-id="how-to-copy-files-to-and-from-the-cluster"><span class="header-section-number">5.4.7</span> How to copy files to and from the cluster</h3>
<p>You may need to transfer files back and forth between your own machine and the cluster. To copy a file called <code>file</code> in a directory called <code>dir</code> on the cluster to the current folder on your own machine, you can use the <code>scp</code> command:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scp</span> <span class="op">&lt;</span>cluster_user_name<span class="op">&gt;</span>@login.genome.au.dk:dir/file .</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>To copy a file called <code>file</code> in the current folder on your own machine to a folder called <code>dir</code> on the cluster, you do this:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scp</span> ./file <span class="op">&lt;</span>cluster_user_name<span class="op">&gt;</span>@login.genome.au.dk:dir/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../notebooks/02_notebook.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Some title</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../howtos/reproducible.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Reproducible research</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>